### PR TITLE
Refactor bad token detector into trait and module

### PR DESCRIPTION
**Pull request by vkgnosis**
I plan on using the trait to implement wrapping detectors that can add a
cache and explicit allow and deny lists.

I dropped the gas price from TokenQuality because I am not sure yet how it should work with the allow and deny lists. Can add this back later.

### Test Plan
not needed

### CODE REVIEW

File: orderbook/src/bad_token/trace_call.rs [hunk 1]
```diff
@@ -11,50 +12,42 @@ use ethcontract::{
...
 /// the amm pair providers.
 /// Tokens are bad if:
 /// - we cannot find an amm pool of the token to one of the base tokens
 /// - transfer into the settlement contract or back out fails
 /// - a transfer loses total balance
-pub struct BadTokenDetector {
+pub struct TraceCallDetector {
     pub web3: Web3,
-    pub pools: Vec<Box<dyn AmmPairProvider>>,
+    pub pools: Vec<Arc<dyn AmmPairProvider>>,
```
Comment (vkgnosis) [inline orderbook/src/bad_token/trace_call.rs:30]: Not needed in this PR but snuck in from another.

File: orderbook/src/bad_token/mod.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+use primitive_types::H160;
+
+/// How well behaved a token is.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum TokenQuality {
+    Good,
+    Bad { reason: String },
+}
+
+impl TokenQuality {
```
Comment (bh2smith) [inline orderbook/src/bad_token/mod.rs:13]: This is cute! Excited to see it in action!

**Review by bh2smith — APPROVED** [review]
Looks excellent. I only wonder if we can expect the logs for bad token quality to come along with the token address as well. I presume this to be true, but don't really see a sign of it here.

**Comment by vkgnosis** [comment]
> Looks excellent. I only wonder if we can expect the logs for bad token quality to come along with the token address as well. I presume this to be true, but don't really see a sign of it here.

Not sure exactly what you mean. This is just some infrastructure for detecting bad tokens. The code that calls it gives it a token address to check so it already knows what the address is and it can do logging how it wants to.

**Review by fleupold — APPROVED** [review]
I like this trait and the offered implementations a lot. All of them are nice little unit testable components that are easy to understand in itself.

**Pull request merged**